lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp

```diff
@@ -151,36 +174,91 @@ static bool optimizeGuaranteedArgument(SILArgument *Arg,
 
 namespace {
 
+/// Stashes a value for an iterator for a scope. This instruction/any
+/// instructions after instruction in the block/any instructions in
+/// other blocks can all be deleted safely.
+struct ForwardInstIteratorGuard {
```

Comment:
[style] This is a substantial abstraction that I would only ever use to hack
around old code instead of just fixing those APIs.

I generally think that transformation utilities that delete instructions (or blocks)
should be responsible for returning a valid iterator. This is just a
style comment...
```
for (auto ii = begin(), nextI = ii, ie = end(); ii != ie; ii = nextI) {
  nextI = std::next(ii)

  // Ususally some large number of conditions...
  if (someBailout)
    continue

  // And any number of possible transformations...
  if (canTransform) {
    nextI = transformInst(ii);
    continue
  }
}
```